src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -379,6 +379,13 @@ void ImageReplayer<I>::handle_bootstrap(int r) {
     m_bootstrap_request = nullptr;
   }
 
+  // set m_delete_requested early to ensure that in case remote
+  // image no longer exists local image gets deleted even if start
+  // is canceled
+  if (r == -ENOLINK) {
+    dout(5) << "remote image no longer exists" << dendl;
+    m_delete_requested = true;
+  }
   if (on_start_interrupted()) {
```

Comment:
Did you see a similar check in `ImageReplayer::bootstrap()`?  If it fires, we wouldn't attempt bootstrapping with the same net effect of local image not getting deleted.

Reply:
I've refreshed the PR after addressing the first 2 comments and I will look into the third one.

Reply:
Removing the check will be sufficient as the bootstrap will return in call to BootstrapRequest::handle_prepare_remote_image() if the remote image does not exist.

Reply:
This check was added in https://github.com/ceph/ceph/commit/91b8a24ea58aee01ec8567195b5aabe11f74b87f for a reason.  By _just_ removing it, you are likely reintroducing the race condition that it took care of.

Reply:
What is the meaning of "the bootstrap is paused behind an image sync." ?

Reply:
> What is the meaning of "the bootstrap is paused behind an image sync." ?

I think it means that a canceled image replayer would still attempt to sync the image as part of bootstrapping (if it sees that it's needed).  This operation can take a while because it's linear in the size of the image and therefore not desired on a canceled image replayer.

> The test added in the referenced commit is for journal mirroring. The current mock tests pass for this fix.

`TestMockImageReplayer.StopBeforeBootstrap` might be imprecise/testing something slightly different.  Does the code flow make it possible for the sync to be kicked off in `BootstrapRequest::image_sync()` with this check removed?

Reply:
In `BootstrapRequest::handle_prepare_remote_image()`, I see that [finish(ENOLINK)](https://github.com/ceph/ceph/blob/fd62a8dfcb345fabed3eea6b3a461f1ca848a239/src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc#L197) called when remote image is missing. Looking at BootstrapRequest state machine diagram,  image_sync happens after preparing the remote image. So maybe removing the check `if (on_start_interrupted(m_lock))`  in `ImageReplayer::bootstrap()` is okay?

Reply:
`BootstrapRequest` is its own class, can't call from `ImageReplayer::on_start_interrupted()` from there.

But we could pass whether we are interrupted to `BootstrapRequest` on construction and then bail after the disposition of local and remote images is determined (and before proceeding with anything else, like image sync).